### PR TITLE
add items to dynamic watch in helm

### LINKED ISSUE #514: helm provisioner doesn't dynamically reconcile underlying bundle contents
When deploying rukpak using the main branch tags, it looks like the helm provisioner doesn't have the same dynamic watch capabilities like the plain provisioner does.

Create the following BD resource:

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleDeployment
metadata:
  name: ahoy-tmp
spec:
  provisionerClassName: core-rukpak-io-helm
  template:
    metadata:
      labels:
        app.kubernetes.io/name: ahoy
    spec:
      provisionerClassName: core-rukpak-io-helm
      source:
        http:
          url: https://github.com/helm/examples/releases/download/hello-world-0.1.0/hello-world-0.1.0.tgz
        type: http
```

And then query for the underlying deployment to verify the helm chart installation succeeded:

```bash
$ k get deployments -n rukpak-system ahoy-tmp-hello-world 
NAME                   READY   UP-TO-DATE   AVAILABLE   AGE
ahoy-tmp-hello-world   1/1     1            1           8s
```

And then delete that resource and verify it doesn't get recreated:

```bash
$ k delete deployments -n rukpak-system ahoy-tmp-hello-world 
deployment.apps "ahoy-tmp-hello-world" deleted
$ k get deployments -w -n rukpak-system
NAME               READY   UP-TO-DATE   AVAILABLE   AGE
core               1/1     1            1           34m
helm-provisioner   1/1     1            1           34m
rukpak-webhooks    1/1     1            1           34m
```

### PR DESCRIPTION
**Pull request by akihikokuroda**
Closes #514

Signed-off-by: akihikokuroda <akuroda@us.ibm.com>

### CODE REVIEW

**Review by joelanford — COMMENTED** [review]
This seems to take a different approach than the plain provisioner. Should this with how the plain provisioner works? That would hopefully make it easier to find re-use opportunities in the future, and it would ensure the UX is generally the same for both.

**Comment by akihikokuroda** [comment]
@joelanford The UX should be the same.  It is doing the same thing as the plain provisioner but it can not do the same way.  It is adding to the watch using the release.manifest after the install because the yaml files in the charts can not be decoded without the rendering.  It is adding the labels using the postrenderer because it must be done against the rendered objects.

**Comment by joelanford** [comment]
Oh right. In Plain, we know the full content of the "templates" ahead of time. In Helm, we have to run the chart through the helm libraries to know what the resulting manifest is.

One other thing that comes to mind. We use a post-renderer under the hood in the helm-operator-plugins repo to inject owner references. If we override the post-renderer to set these labels, we might be overriding the ownerref post renderer.

**Comment by akihikokuroda** [comment]
> If we override the post-renderer to set these labels, we might be overriding the ownerref post renderer.

Yes, I found that.  I added the ownerref injection in the post-render function.

**Comment by joelanford** [comment]
Not blocking for this PR, but I imagine that unknowingly setting a post-renderer like you did could be a footgun for others. I wonder if there's a way in helm-operator-plugins to support adding more caller-defined post-renderers to the chain.

**Comment by akihikokuroda** [comment]
I made changes to call the preset PostRenderer after our PostRenderer if it is set.

**Comment by joelanford** [comment]
@akihikokuroda I've got a PR up in helm-operator-plugins that I think should get rid of the awkward hacking you had to do to use both the built-in and extra post-renderer.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (timflannagan) [inline internal/provisioner/helm/controllers/bundledeployment_controller.go:400]: Is printing the obj here potentially problem for logs? Or does this just print the metadata information we care about?

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (timflannagan) [inline internal/provisioner/helm/controllers/bundledeployment_controller.go:326]: We may want/need a tech debt ticket here given the number of parameters for this method.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (timflannagan) [inline internal/provisioner/helm/controllers/bundledeployment_controller.go:347]: Any idea on why this comparison is needed?

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (timflannagan) [inline internal/provisioner/helm/controllers/bundledeployment_controller.go:281]: This feels like something the plain provisioner could also hook into?

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (timflannagan) [inline internal/provisioner/helm/controllers/bundledeployment_controller.go:194]: nit: I think we have a const variable somewhere for the BD GVK.

**Review by timflannagan — COMMENTED** [review]
These changes seem reasonable. If we're hacking around the helm-operator-plugins implementation here, does it make sense to create a ticket that tracks cleaning up the implementation once Joe's changes land and a new release is cut?

**Comment by akihikokuroda** [comment]
@timflannagan I created the issue for refactoring. https://github.com/operator-framework/rukpak/issues/534

**Comment by timflannagan** [comment]
Awesome. Follow-up question: do we think it would be valuable to also add a comment linking to that issue for the postrenderer hacks?

**Comment by akihikokuroda** [comment]
That's a good idea.  I'll do it.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (akihikokuroda) [reply internal/provisioner/helm/controllers/bundledeployment_controller.go:194]: It doesn't seem defined.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (akihikokuroda) [reply internal/provisioner/helm/controllers/bundledeployment_controller.go:281]: This is checking for the changes in Chart.yaml file.  The plain provisioner is generating the contents of the Chart.yam file on the fly so it's always same.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (akihikokuroda) [reply internal/provisioner/helm/controllers/bundledeployment_controller.go:347]: This is checking for the changes in Chart.yaml file.  It's contents are [here](https://pkg.go.dev/helm.sh/helm/v3@v3.9.4/pkg/chart#Metadata).

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (akihikokuroda) [reply internal/provisioner/helm/controllers/bundledeployment_controller.go:326]: OK.  I'll add a TODO comment.

File: internal/provisioner/helm/controllers/bundledeployment_controller.go
Comment (akihikokuroda) [reply internal/provisioner/helm/controllers/bundledeployment_controller.go:400]: The `err` has enough information.  I take out the `obj`.

**Comment by akihikokuroda** [comment]
Refactor has done.  Please review.

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (timflannagan) [inline internal/provisioner/bundledeployment/bundledeployment.go:507]: I haven't had a chance to look over the new directory structure since #447 landed: we're specifying the plain.ProvisionerID here, and filtering BundleDeployments that don't match that unique provisioner class ID from the cache. Is that the right read here? If so, I think we'll need to solve this in an agnostic way s.t. our core provisioners can leverage this functionality without inter-depending on the plain provisioner libraries.

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (timflannagan) [inline internal/provisioner/bundledeployment/bundledeployment.go:454]: nit: can we copy/paste the 534 issue link so it's easier to track when reading through the codebase locally?

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (timflannagan) [inline internal/provisioner/bundledeployment/bundledeployment.go:274]: nit: we have a constant variable for the BundleDeployment Kind: https://github.com/operator-framework/rukpak/blob/main/api/v1alpha1/bundledeployment_types.go#L26.

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (akihikokuroda) [reply internal/provisioner/bundledeployment/bundledeployment.go:507]: Oh, it's my mistake.  This function is not used any more since #447 changes.   I'll take this out.

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (akihikokuroda) [reply internal/provisioner/bundledeployment/bundledeployment.go:454]: Sure.  It must be better.

File: internal/provisioner/bundledeployment/bundledeployment.go
Comment (akihikokuroda) [reply internal/provisioner/bundledeployment/bundledeployment.go:274]: Thanks!  I'll update.

**Comment by timflannagan** [comment]
Sorry for the slow turnaround on this. I've been pretty heads down with other projects/work recently. Can we also re-title this PR to fix the typo?

**Comment by timflannagan** [comment]
Pulled this down and quickly testing it out.

**Review by timflannagan — APPROVED** [review]
I tested these changes out locally, and it looks like we're good to go:

<details>

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleDeployment
metadata:
  name: hello-world-helm
  namespace: demo1
spec:
  provisionerClassName: core-rukpak-io-helm
  template:
    metadata:
      labels:
        app: hello-world-helm
    spec:
      provisionerClassName: core-rukpak-io-helm
      source:
        http:
          url: https://github.com/helm/examples/releases/download/hello-world-0.1.0/hello-world-0.1.0.tgz
        type: http
```

</details>

And then deleted the deployment that got created from that Bundle resource a couple of times, and waited for it to be quickly recreated.

I haven't had a chance to catch up the new repository structure, but this looks like a reasonable short term hack while we wait for a release to be cut.

**Review by tylerslaton — APPROVED** [review]

**Pull request merged**
